File: workers/temporal-worker/src/safety/budgets.ts
/**
 * Client for interacting with the budget tracking API.
 */

import pino from 'pino';

const logger = pino();

/**
 * Status of a budget for a specific run.
 */
export interface BudgetStatus {
  run_id: string;
  spent: number;
  limit: number;
  remaining: number;
  exceeded: boolean;
}

/**
 * Request to create a new budget.
 */
export interface CreateBudgetRequest {
  run_id: string;
  limit: number;
}

/**
 * Request to record spending.
 */
export interface SpendBudgetRequest {
  run_id: string;
  amount: number;
}

/**
 * Request to check if spending is allowed.
 */
export interface CanSpendRequest {
  run_id: string;
  amount: number;
}

/**
 * Response indicating if spending is allowed.
 */
export interface CanSpendResponse {
  can_spend: boolean;
  current_status: BudgetStatus;
}

/**
 * Client for tracking and enforcing spending budgets.
 */
export class BudgetClient {
  private readonly baseUrl: string;

  /**
   * Create a new budget client.
   *
   * @param baseUrl - Base URL of the FastAPI service (default: http://localhost:8000)
   */
  constructor(baseUrl: string = 'http://localhost:8000') {
    this.baseUrl = baseUrl.replace(/\/$/, ''); // Remove trailing slash
  }

  /**
   * Create a new budget for a run.
   *
   * @param runId - Unique identifier for the run
   * @param limit - Budget limit in USD
   * @returns Initial budget status
   */
  async createBudget(runId: string, limit: number): Promise<BudgetStatus> {
    logger.info({ runId, limit }, 'Creating budget');

    try {
      const response = await fetch(`${this.baseUrl}/api/v1/safety/budget/create`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          run_id: runId,
          limit,
        } as CreateBudgetRequest),
      });

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`Failed to create budget: ${response.status} ${error}`);
      }

      const status = await response.json() as BudgetStatus;
      logger.info({ runId, status }, 'Budget created successfully');
      return status;

    } catch (error) {
      logger.error({ error, runId, limit }, 'Failed to create budget');
      throw new Error(
        `Failed to create budget: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }

  /**
   * Record spending against a budget.
   *
   * @param runId - Unique identifier for the run
   * @param amount - Amount to spend in USD
   * @returns Updated budget status
   */
  async spend(runId: string, amount: number): Promise<BudgetStatus> {
    logger.info({ runId, amount }, 'Recording spend');

    try {
      const response = await fetch(`${this.baseUrl}/api/v1/safety/budget/spend`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          run_id: runId,
          amount,
        } as SpendBudgetRequest),
      });

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`Failed to record spend: ${response.status} ${error}`);
      }

      const status = await response.json() as BudgetStatus;

      if (status.exceeded) {
        logger.warn({ runId, status }, 'Budget exceeded');
      }

      return status;

    } catch (error) {
      logger.error({ error, runId, amount }, 'Failed to record spend');
      throw new Error(
        `Failed to record spend: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }

  /**
   * Get current budget status for a run.
   *
   * @param runId - Unique identifier for the run
   * @returns Current budget status
   */
  async getStatus(runId: string): Promise<BudgetStatus> {
    logger.info({ runId }, 'Getting budget status');

    try {
      const response = await fetch(`${this.baseUrl}/api/v1/safety/budget/${runId}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      });

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`Failed to get budget status: ${response.status} ${error}`);
      }

      const status = await response.json() as BudgetStatus;
      return status;

    } catch (error) {
      logger.error({ error, runId }, 'Failed to get budget status');
      throw new Error(
        `Failed to get budget status: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }

  /**
   * Check if a spend amount would exceed the budget.
   *
   * @param runId - Unique identifier for the run
   * @param amount - Amount to check in USD
   * @returns True if the spend would be within budget
   */
  async canSpend(runId: string, amount: number): Promise<boolean> {
    logger.info({ runId, amount }, 'Checking can spend');

    try {
      const response = await fetch(`${this.baseUrl}/api/v1/safety/budget/can-spend`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          run_id: runId,
          amount,
        } as CanSpendRequest),
      });

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`Failed to check can spend: ${response.status} ${error}`);
      }

      const result = await response.json() as CanSpendResponse;
      return result.can_spend;

    } catch (error) {
      logger.error({ error, runId, amount }, 'Failed to check can spend');
      throw new Error(
        `Failed to check can spend: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }

  /**
   * Delete a budget and all associated data.
   *
   * @param runId - Unique identifier for the run
   */
  async deleteBudget(runId: string): Promise<void> {
    logger.info({ runId }, 'Deleting budget');

    try {
      const response = await fetch(`${this.baseUrl}/api/v1/safety/budget/${runId}`, {
        method: 'DELETE',
        headers: {
          'Content-Type': 'application/json',
        },
      });

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`Failed to delete budget: ${response.status} ${error}`);
      }

      logger.info({ runId }, 'Budget deleted successfully');

    } catch (error) {
      logger.error({ error, runId }, 'Failed to delete budget');
      throw new Error(
        `Failed to delete budget: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }

  /**
   * Check if budget is exceeded.
   *
   * @param runId - Unique identifier for the run
   * @returns True if budget is exceeded
   */
  async isExceeded(runId: string): Promise<boolean> {
    try {
      const status = await this.getStatus(runId);
      return status.exceeded;
    } catch {
      return false;
    }
  }

  /**
   * Enforce budget limit before spending.
   * Throws an error if spending would exceed budget.
   *
   * @param runId - Unique identifier for the run
   * @param amount - Amount to spend
   * @throws Error if spending would exceed budget
   */
  async enforceLimit(runId: string, amount: number): Promise<void> {
    const canSpend = await this.canSpend(runId, amount);

    if (!canSpend) {
      const status = await this.getStatus(runId);
      throw new Error(
        `Budget limit exceeded: Would spend $${amount}, but only $${status.remaining} remaining of $${status.limit} limit`
      );
    }
  }

  /**
   * Spend with enforcement.
   * Checks budget before spending and throws if exceeded.
   *
   * @param runId - Unique identifier for the run
   * @param amount - Amount to spend
   * @returns Updated budget status
   * @throws Error if spending would exceed budget
   */
  async spendWithEnforcement(runId: string, amount: number): Promise<BudgetStatus> {
    await this.enforceLimit(runId, amount);
    return await this.spend(runId, amount);
  }
}

File: apps/api/ae_api/safety/budgets.py
"""Budget tracking and enforcement for cost control."""


import structlog
from pydantic import BaseModel, Field
from redis.asyncio import Redis

logger = structlog.get_logger()


class BudgetStatus(BaseModel):
    """Status of a budget for a specific run."""

    run_id: str = Field(description="Unique identifier for the run")
    spent: float = Field(description="Amount spent so far in USD")
    limit: float = Field(description="Budget limit in USD")
    remaining: float = Field(description="Remaining budget in USD")
    exceeded: bool = Field(description="Whether budget has been exceeded")


class BudgetTracker:
    """Tracks and enforces spending budgets using Redis."""

    BUDGET_KEY_PREFIX = "budget:"
    SPENT_KEY_PREFIX = "spent:"
    EXCEEDED_KEY_PREFIX = "exceeded:"
    DEFAULT_TTL = 86400 * 7  # 7 days in seconds

    def __init__(self, redis_client: Redis):
        """
        Initialize budget tracker with Redis client.

        Args:
            redis_client: Redis client for tracking budget state
        """
        self.redis = redis_client

    async def create_budget(self, run_id: str, limit: float) -> BudgetStatus:
        """
        Create a new budget for a run.

        Args:
            run_id: Unique identifier for the run
            limit: Budget limit in USD

        Returns:
            Initial budget status

        Raises:
            ValueError: If budget limit is invalid
        """
        if limit <= 0:
            raise ValueError(f"Budget limit must be positive, got {limit}")

        logger.info("Creating budget", run_id=run_id, limit=limit)

        # Set budget limit
        budget_key = f"{self.BUDGET_KEY_PREFIX}{run_id}"
        await self.redis.set(budget_key, str(limit), ex=self.DEFAULT_TTL)

        # Initialize spent to 0
        spent_key = f"{self.SPENT_KEY_PREFIX}{run_id}"
        await self.redis.set(spent_key, "0", ex=self.DEFAULT_TTL)

        # Set exceeded flag to false
        exceeded_key = f"{self.EXCEEDED_KEY_PREFIX}{run_id}"
        await self.redis.set(exceeded_key, "0", ex=self.DEFAULT_TTL)

        return BudgetStatus(
            run_id=run_id,
            spent=0.0,
            limit=limit,
            remaining=limit,
            exceeded=False,
        )

    async def spend(self, run_id: str, amount: float) -> BudgetStatus:
        """
        Record spending against a budget.

        Args:
            run_id: Unique identifier for the run
            amount: Amount to spend in USD

        Returns:
            Updated budget status

        Raises:
            ValueError: If amount is negative or budget doesn't exist
        """
        if amount < 0:
            raise ValueError(f"Amount must be non-negative, got {amount}")

        logger.info("Recording spend", run_id=run_id, amount=amount)

        # Check if budget exists
        budget_key = f"{self.BUDGET_KEY_PREFIX}{run_id}"
        limit_str = await self.redis.get(budget_key)

        if limit_str is None:
            raise ValueError(f"Budget not found for run_id: {run_id}")

        limit = float(limit_str)

        # Increment spent amount atomically
        spent_key = f"{self.SPENT_KEY_PREFIX}{run_id}"
        spent = float(await self.redis.incrbyfloat(spent_key, amount))

        # Check if budget exceeded
        exceeded = spent > limit
        remaining = max(0.0, limit - spent)

        # Update exceeded flag if necessary
        if exceeded:
            exceeded_key = f"{self.EXCEEDED_KEY_PREFIX}{run_id}"
            await self.redis.set(exceeded_key, "1")
            logger.warning(
                "Budget exceeded",
                run_id=run_id,
                spent=spent,
                limit=limit,
                exceeded_by=spent - limit,
            )

        return BudgetStatus(
            run_id=run_id,
            spent=spent,
            limit=limit,
            remaining=remaining,
            exceeded=exceeded,
        )

    async def get_status(self, run_id: str) -> BudgetStatus:
        """
        Get current budget status for a run.

        Args:
            run_id: Unique identifier for the run

        Returns:
            Current budget status

        Raises:
            ValueError: If budget doesn't exist
        """
        budget_key = f"{self.BUDGET_KEY_PREFIX}{run_id}"
        spent_key = f"{self.SPENT_KEY_PREFIX}{run_id}"
        exceeded_key = f"{self.EXCEEDED_KEY_PREFIX}{run_id}"

        # Get all values with pipeline for efficiency
        async with self.redis.pipeline() as pipe:
            pipe.get(budget_key)
            pipe.get(spent_key)
            pipe.get(exceeded_key)
            results = await pipe.execute()

        limit_str, spent_str, exceeded_str = results

        if limit_str is None:
            raise ValueError(f"Budget not found for run_id: {run_id}")

        limit = float(limit_str)
        spent = float(spent_str or "0")
        exceeded = exceeded_str == b"1" if isinstance(exceeded_str, bytes) else exceeded_str == "1"
        remaining = max(0.0, limit - spent)

        return BudgetStatus(
            run_id=run_id,
            spent=spent,
            limit=limit,
            remaining=remaining,
            exceeded=exceeded,
        )

    async def check_can_spend(self, run_id: str, amount: float) -> bool:
        """
        Check if a spend amount would exceed the budget.

        Args:
            run_id: Unique identifier for the run
            amount: Amount to check in USD

        Returns:
            True if the spend would be within budget
        """
        try:
            status = await self.get_status(run_id)

            # If already exceeded, no more spending allowed
            if status.exceeded:
                logger.warning("Budget already exceeded", run_id=run_id)
                return False

            # Check if this spend would exceed budget
            would_exceed = (status.spent + amount) > status.limit

            if would_exceed:
                logger.warning(
                    "Spend would exceed budget",
                    run_id=run_id,
                    amount=amount,
                    spent=status.spent,
                    limit=status.limit,
                )

            return not would_exceed

        except ValueError as e:
            logger.error("Error checking budget", run_id=run_id, error=str(e))
            return False

    async def delete_budget(self, run_id: str) -> None:
        """
        Delete a budget and all associated data.

        Args:
            run_id: Unique identifier for the run
        """
        logger.info("Deleting budget", run_id=run_id)

        budget_key = f"{self.BUDGET_KEY_PREFIX}{run_id}"
        spent_key = f"{self.SPENT_KEY_PREFIX}{run_id}"
        exceeded_key = f"{self.EXCEEDED_KEY_PREFIX}{run_id}"

        await self.redis.delete(budget_key, spent_key, exceeded_key)
